File: advent_of_code/2019/day1/day1.py
fileName = "day1.in"
data = map(lambda s: int(s.replace('\n', '')), open(fileName, 'r+').readlines())

# Part 1
import math

def get_fuel(mass):
    return math.floor(mass / 3) - 2

print(sum(map(get_fuel, data)))

# Part 2
def get_fuel_recur(mass, acc=0):
    result = get_fuel(mass)
    if math.floor(result / 3) < 2:
        return result + acc
    else:
        return get_fuel_recur(result, result+acc)

print(sum(map(get_fuel_recur, data)))

File: advent_of_code/2019/day1/day1.scala
import scala.io.Source

val fileName = "day1.in"
val lines = Source.fromFile(fileName).getLines.toList.map(s => s.toInt)

// Part 1

def getFuel(mass: Int) : Int = {
  (mass / 3.0).floor.toInt - 2
}
println(lines.map(getFuel).sum)

// Part 2
def getFuel(mass: Int, acc: Int) : Int = {
  val result = (mass / 3.0).floor.toInt - 2
  if ((result / 3.0).floor.toInt < 2) {
    result + acc
  } else (
    getFuel(result, result + acc)
  )
}

def wrapper(mass: Int) : Int = {
  getFuel(mass, 0)
}

println(lines.map(wrapper).sum)
